source/common/http/http1/codec_impl.cc

```diff
@@ -687,7 +687,13 @@ Envoy::StatusOr<size_t> ConnectionImpl::dispatchSlice(const char* slice, size_t
 
   const ParserStatus status = parser_->getStatus();
   if (status != ParserStatus::Ok && status != ParserStatus::Paused) {
-    RETURN_IF_ERROR(sendProtocolError(Http1ResponseCodeDetails::get().HttpCodecError));
+    absl::string_view error = Http1ResponseCodeDetails::get().HttpCodecError;
+    if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.http1_use_balsa_parser") &&
+        (parser_->errorMessage() == "headers size exceeds limit" ||
+         parser_->errorMessage() == "trailers size exceeds limit")) {
```

Comment:
optional: What do you think of changing "headers size exceeds limit" and "trailers size exceeds limit" to Http1ResponseCodeDetails::get().HeadersTooLarge in balsa_parser.cc? Then here it can be:

if (feature_enabled && parser_->errorMessage() == HeadersTooLarge) {
  error = HeadersTooLarge;
}

Which reads just a tiny bit simpler here, even if we're coupling a bit more tightly with balsa_parser this way.

If wanting to keep as-is, maybe could define "headers size exceeds limit" and "trailers size exceeds limit" as constants for reuse here and in balsa_parser.cc.

Reply:
Unfortunately there are two different kind of error messages, see, for example:
https://sourcegraph.com/github.com/envoyproxy/envoy@a00c314/-/blob/test/common/http/http1/codec_impl_test.cc?L2999-3000

My current goal with BalsaParser is feature parity with http-parser, so I'm crafting each error to match.

For reference, when using http-parser, it is not http-parser itself emitting the error, but ConnectionImpl synthesizing it at https://sourcegraph.com/github.com/envoyproxy/envoy@a00c314/-/blob/source/common/http/http1/codec_impl.cc?L551-555, the two kinds of error at the same time.

However, when using BalsaParser, BalsaFrame emits a BalsaFrameEnums::ErrorCode, which is not surfaced directly to ConnectionImpl, but instead converted into an error message string in BalsaParser::HandleError() to match status.message().  Then ConnectionImpl has to synthesize the other kind of error message.

The reason BalsaParser must enforce the size limit internally is that it only calls back to ConnectionImpl when a header key-value pair is parsed completely.  Since we do not want BalsaParser to do unlimited buffering, I'm setting its internal limit to the same as ConnectionImpl's at https://sourcegraph.com/github.com/envoyproxy/envoy@a00c314/-/blob/source/common/http/http1/balsa_parser.cc?L97.  If it wasn't for this complication, we could just rely on the same code path as http-parser does to emit the appropriate error messages.

One option is to surface BalsaFrameEnums::ErrorCode, and move conversion to string logic to ConnectionImpl.  Then ConnectionImpl could set both error message strings based on an enum, not setting one string based on the other.  Note that what you're looking at in this PR is the special case: in all other cases, the responseDetails() kind of error message is HttpCoderError, emitted by ConnectionImpl::dispatchSlice().

Another option is to move this logic to BalsaParser, and let it emit two kinds of error messages: one can be HttpCoderError except when it needs to be HeadersTooLarge, the other BalsaFrameEnums::ErrorCodeToString(error_code) except when http-parser-matching error string is required.

I would prefer to land this PR as-is, and revisit this when all error cases are covered.  A reasonable approach could be changing the behavior so that the error is always HttpCoderError, even when currently it is HeadersTooLarge.  Or a more bold move would be to make status.message() always match responseDetails().  In any case, some cleanup will be in order, because BalsaParser currently emits a number of all-uppercase error messages with HPE_ prefix, which is http-parser specific.  This I plan to do after the BalsaParser transition is done, guarded by a different runtime flag if necessary.  That might be a good time to clean up error generating logic as well.

Reply:
Looks like the error message is constructed here: https://sourcegraph.com/github.com/envoyproxy/envoy@a00c314b8a004d1519a464089fc5e0dfc41547fa/-/blob/source/common/http/http1/codec_impl.cc?L589

The changes make sense to me.

Reply:
Thanks for the in-depth explanation of your response! Thanks also for the pointers to the different error messages, interesting. I see a lot of careful crafting to maintain parity with http-parser :D: 

Full support on landing this PR as-is and thinking about revisiting this for clean-up when all error cases are covered.

Reply:
SGTM, can you track this in the issue for integrating with Balsa, so we do end up cleaning this up? Thanks 
> I would prefer to land this PR as-is, and revisit this when all error cases are covered.

Reply:
Thanks.  I added comment https://github.com/envoyproxy/envoy/issues/21245#issuecomment-1209476687 for this.  Let me know if you think this needs to be tracked in a more prominent way.